File: src/ui/content/editNodes.ts
import {
  Id, StringSlice, Sentence, TreeAndNodeId
} from '../../content/types';
import { InsertedNode, transformAllNodesInTree } from '../../content/unpositioned/manipulation';
import { isTerminal, UnpositionedNode, UnpositionedTree } from '../../content/unpositioned/types';

export type NodeSelectionInPlot = { nodes: TreeAndNodeId[] };
export type SliceSelectionInPlot = { treeId: Id, slice: StringSlice };
export type SelectionInPlot = NodeSelectionInPlot | SliceSelectionInPlot;

const isWordChar = (char: string) => /['A-Za-z\u00c0-\u1fff]/.test(char);

/**
 * Returns the range of the word that the given position is within.
 */
const getWordRange = (sentence: Sentence, position: number): [start: number, end: number] => {
  let start = position;
  while (start > 0 && isWordChar(sentence[start - 1])) start--;

  let end = position;
  while (end < sentence.length && isWordChar(sentence[end])) end++;

  return [start, end];
};

/**
 * Returns the definition for a new node based on the current selection.
 * If the selection is a slice of length 0, expands it to span the word where the cursor currently is,
 * using the given sentence.
 */
export const newNodeFromSelection = (selection: SelectionInPlot, sentence: Sentence): InsertedNode => {
  // A slice of the sentence is selected
  if ('slice' in selection) {
    const [sliceStart, sliceEnd] = selection.slice;
    const selectedSlice = sliceStart !== sliceEnd
      ? selection.slice
      : getWordRange(sentence, sliceStart);
    return {
      targetSlice: selectedSlice,
      label: '',
    };
  }
  // One or more nodes are selected
  const selectedNodeIds = selection.nodes.map(({ nodeId }) => nodeId);
  return {
    targetChildIds: selectedNodeIds,
    label: '',
  };
};

/**
 * Shifts the slice associated with a node by the given number of characters,
 * based on the kind of selection made and the operation done on it.
 * Node slices will not expand when adding characters, but may contract when removing characters.
 */
const shiftNodeSliceAfterChange =
  (oldSelection: StringSlice, shiftBy: number) =>
  (node: UnpositionedNode): UnpositionedNode => {
    if (                               // If:
      !isTerminal(node) ||             // this is not a terminal node, or
      shiftBy === 0 ||                 // the sentence length did not change, or
      node.slice[1] < oldSelection[0]  // the selection was entirely after the node, then
    ) return node;                     // no change is necessary

    const [oldNodeSliceStart, oldNodeSliceEnd] = node.slice;
    const newNodeSlice: StringSlice = [oldNodeSliceStart, oldNodeSliceEnd];
    if (oldSelection[0] === oldSelection[1]) {  // No selection, just a cursor
      const oldCursorPos = oldSelection[0];
      if (oldCursorPos === oldNodeSliceStart) {  // Cursor was at the beginning of the slice
        newNodeSlice[1] += shiftBy;
        if (shiftBy > 0) {  // If adding, move the slice forward; if removing, contract it
          newNodeSlice[0] += shiftBy;
        }
      }
      if (oldCursorPos === oldNodeSliceEnd) {  // Cursor was at the end of the slice
        if (shiftBy < 0) {  // If removing, contract the slice
          newNodeSlice[1] += shiftBy;
        }
      }
      if (oldCursorPos < oldNodeSliceStart) {  // Cursor was before the slice
        newNodeSlice[0] += shiftBy;
        newNodeSlice[1] += shiftBy;
      }
      if (oldNodeSliceStart < oldCursorPos && oldCursorPos < oldNodeSliceEnd) {  // Cursor was in the middle of the slice
        newNodeSlice[1] += shiftBy;
      }
    } else {  // A selection spanning at least one character was made
      console.error('Not implemented yet');
    }
    return { ...node, slice: newNodeSlice };
  }

/**
 * Responds to a change in the sentence associated with the given tree,
 * adapting existing nodes to the change if any exist.
 * @param newSentence The new sentence to use for the tree.
 * @param oldSelection The selection before the change was made.
 *   This is used to determine how exactly node ranges should change.
 */
export const handleLocalSentenceChange =
  (newSentence: Sentence, oldSelection: StringSlice) =>
  (tree: UnpositionedTree): UnpositionedTree => transformAllNodesInTree(
    shiftNodeSliceAfterChange(oldSelection, newSentence.length - tree.sentence.length)
  )({
    ...tree,
    sentence: newSentence,
  });

File: npbloom-core/src/commonMain/kotlin/space/yuvalinguist/npbloom/ui/content/EditNodes.kt
package space.yuvalinguist.npbloom.ui.content

import space.yuvalinguist.npbloom.content.Id
import space.yuvalinguist.npbloom.content.Sentence
import space.yuvalinguist.npbloom.content.StringSlice
import space.yuvalinguist.npbloom.content.unpositioned.*
import space.yuvalinguist.npbloom.ui.NodeSelectionInPlot
import space.yuvalinguist.npbloom.ui.SelectionInPlot
import space.yuvalinguist.npbloom.ui.SliceSelectionInPlot

private fun isWordChar(char: Char) = char == '\'' || (char.isLetterOrDigit())

/**
 * Returns the range of the word that the given position is within.
 */
private fun getWordRange(sentence: Sentence, position: Int): StringSlice {
    var start = position
    while (start > 0 && isWordChar(sentence[start - 1])) start--

    var end = position
    while (end < sentence.length && isWordChar(sentence[end])) end++

    return StringSlice(start, end)
}

/**
 * Spreads a zero-length slice to cover the entire word.
 */
fun spreadSlice(slice: StringSlice, sentence: Sentence) =
    if (slice.isZeroLength) getWordRange(sentence, slice.start) else slice

/**
 * Returns the definition for a new node based on the current selection.
 * If the selection is a slice of length 0, expands it to span the word where the cursor currently is,
 * using the given sentence.
 */
internal fun newNodeFromSelection(newNodeId: Id, selection: SelectionInPlot, sentence: Sentence): InsertedNode =
    when (selection) {
        is SliceSelectionInPlot -> {
            // A slice of the sentence is selected
            val sliceAfterSpread = spreadSlice(selection.slice, sentence)
            InsertedTerminalNode(
                newNodeId, "", null, sliceAfterSpread,
                triangle = sliceAfterSpread crossesWordBoundaryIn sentence
            )
        }

        is NodeSelectionInPlot -> {
            // One or more nodes are selected
            val selectedNodeIds = selection.nodeIndicators.map { it.nodeId }.toSet()
            InsertedBranchingNode(newNodeId, "", null, selectedNodeIds)
        }

        else -> error("Cannot create a new node from this selection ($selection)")
    }

/**
 * Shifts and/or resizes the slice by the given number of characters in response to a change in the sentence,
 * based on the kind of selection made and the operation done on it.
 * Node slices will not expand when adding characters, but may contract when removing characters.
 */
private fun StringSlice.shiftAfterChange(oldSelection: StringSlice, shiftBy: Int): StringSlice {
    if (shiftBy == 0) return this

    var (newStart, newEnd) = this
    if (oldSelection.isZeroLength) {  // No selection, just a cursor
        when {
            oldSelection.start == start -> {  // Cursor was at the beginning of the slice
                newEnd += shiftBy
                if (shiftBy > 0) {  // If adding, move the slice forward; if removing, contract it
                    newStart += shiftBy
                }
            }

            oldSelection.start == endExclusive -> {  // Cursor was at the end of the slice
                if (shiftBy < 0) {  // If removing, contract the slice
                    newEnd += shiftBy
                }
            }

            oldSelection.start < start -> {  // Cursor was before the slice
                newStart += shiftBy
                newEnd += shiftBy
            }

            oldSelection.start in (start + 1) until endExclusive -> {  // Cursor was in the middle of the slice
                newEnd += shiftBy
            }
        }
    } else {  // A selection spanning at least one character was made
        when {
            overlapsWith(oldSelection) -> {
                return StringSlice(oldSelection.start, oldSelection.start)
            }

            oldSelection.start < start -> {  // Selection was before the slice
                newStart += shiftBy
                newEnd += shiftBy
            }
        }
    }
    return StringSlice(newStart, newEnd)
}

private fun UnpositionedNode.shiftNodeSliceAfterChange(oldSelection: StringSlice, shiftBy: Int): UnpositionedNode {
    if (                                         // If:
        this !is UnpositionedTerminalNode ||     // this is not a terminal node, or
        shiftBy == 0 ||                          // the sentence length did not change, or
        slice.endExclusive < oldSelection.start  // the selection was entirely after the node, then
    ) return this                                // no change is necessary

    val newSlice = slice.shiftAfterChange(oldSelection, shiftBy)
    if (!oldSelection.isZeroLength && slice overlapsWith oldSelection) {
        return UnpositionedFormerlyTerminalNode(id, label, offset, newSlice, triangle)
    }
    return UnpositionedTerminalNode(id, label, offset, newSlice, triangle)
}

/**
 * Responds to a change in the sentence associated with the given tree,
 * adapting existing nodes to the change if any exist.
 * @param newSentence The new sentence to use for the tree.
 * @param oldSelection The selection before the change was made.
 *   This is used to determine how exactly node ranges should change.
 */
internal fun UnpositionedTree.handleLocalSentenceChange(newSentence: Sentence, oldSelection: StringSlice) =
    UnpositionedTree(id, newSentence, nodes, coordsInPlot).transformAllNodes {
        it.shiftNodeSliceAfterChange(oldSelection, newSentence.length - sentence.length)
    }
